### PR TITLE
vod: New webhooks & simplified asset responses

### LINKED ISSUE #1148: Simpler webhooks for VOD
- rename task.created -> task.started (or spawned)
 - task.status -> task.updated; asset.status -> asset.updated
 - New webhooks:
   - asset.error
   - asset.ready
   - asset.deleted

### LINKED ISSUE #1149: Simplify VOD API responses
- task -> taskId in responses
              - remove some fields from asset (by default, maybe allow via querystring)

### PR DESCRIPTION
**Pull request by gioelecerati**
**What does this pull request do? Explain your changes. (required)**

Rename the `task.created` webhook to `task.spawned`, simplify the asset responses

**Specific updates (required)**

- Generating a signed url now responds only the ID of the task instead of the full task object
- The `/api/asset` endpoint returns the asset object without the videoSpec field
- The same endpoint returns the videoSpec if `details=true` is passed. 
- `task.created` -> `task.spawned` webhook. `createTask` -> `spawnTask` in TaskScheduler
- Implement webhooks as #1148 
- Integrate #1136 
- Change the docs accordingly to the new responses, webhooks and request params

**How did you test each of these updates (required)**

yarn test

**Does this pull request close any open issues?**

Fixes #1148
Fixes #1149 

**Screenshots (optional):**

**Checklist:**

- [X] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/api/src/controllers/asset.ts [hunk 1]
```diff
@@ -55,6 +55,17 @@ function validateAssetMeta(meta: Record<string, string>) {
...
 
+function cleanAssetResponse(assets: WithID<Asset>[]) {
+  let cleanedAssets = [];
+  for (var i = 0; i < assets.length; i++) {
+    if (assets[i].videoSpec) {
+      delete assets[i].videoSpec.tracks;
+    }
+    cleanedAssets.push(assets[i]);
+  }
+  return cleanedAssets;
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: **nit:** Functionally, this could be:
```suggestion
  return assets.map(a => !a.videoSpec ? a : { 
    ...a, 
    videoSpec: {
      ...a.videoSpec,
      tracks: undefined
    }
  })
```
WDYT?

File: packages/api/src/controllers/asset.ts [hunk 2]
```diff
@@ -343,7 +358,7 @@ app.get("/:id", authorizer({}), async (req, res) => {
     );
   }
 
-  res.json(asset);
+  res.json(db.asset.cleanWriteOnlyResponse(asset));
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: This is not necessary as the `mung` JSON middleware already does that.

https://github.com/livepeer/livepeer-com/blob/98652ad082a738f3697e17862192994946786184/packages/api/src/controllers/asset.ts#L212

File: packages/api/src/controllers/asset.ts [hunk 3]
```diff
@@ -328,7 +339,11 @@ app.get("/", authorizer({}), async (req, res) => {
     res.links({ next: makeNextHREF(req, newCursor) });
   }
 
-  return res.json(output);
+  if (details) {
+    return res.json(output);
+  } else {
+    return res.json(cleanAssetResponse(output));
+  }
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: I think this should be done on that same [`mung` middleware](https://github.com/livepeer/livepeer-com/blob/98652ad082a738f3697e17862192994946786184/packages/api/src/controllers/asset.ts#L212). This will allow it to:
 - Clean the asset objects in every API, not only in the list here (we want that, right?)
 - Keep any response-cleaning logic centralized in that function, so there's only one place to look (even if it has some internal helper, there's still only one entry-point for that)

WDYT?

File: packages/api/src/controllers/asset.ts [hunk 4]
```diff
@@ -527,7 +542,7 @@ app.post(
       asset
     );
 
-    res.json({ url, asset, task });
+    res.json({ url, asset, task: { id: task.id } });
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: Hmm I feel like there are other places where we return tasks here. Shouldn't you be updating `transcode`, `request-upload` and `import` as well?

File: packages/api/src/controllers/helpers.ts [hunk 5]
```diff
@@ -108,6 +108,15 @@ function bytesToHexString(bytes: Uint8Array, separate = false) {
   return result;
 }
 
+export function cleanAdminOnlyFields(
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts]: Was this used anywhere new?

File: packages/api/src/schema/schema.yaml [hunk 6]
```diff
@@ -211,10 +211,14 @@ components:
...
               - playback.user.new
               - asset.created
-              - asset.status
-              - task.created
-              - task.status
-              - task.finished
+              - asset.updated
+              - asset.failed
+              - asset.ready
+              - asset.deleted
```
Comment (victorges) [inline packages/api/src/schema/schema.yaml:217]: Need to fire this from `app.deleted("/:id")` in `asset.ts`

File: packages/api/src/schema/schema.yaml [hunk 7]
```diff
@@ -211,10 +211,14 @@ components:
...
+              - task.spawned
+              - task.updated
+              - task.succeeded
```
Comment (victorges) [inline packages/api/src/schema/schema.yaml]: Reminder to change this to `task.completed`

File: packages/api/src/task/scheduler.ts [hunk 8]
```diff
@@ -244,22 +246,27 @@ export default class TaskScheduler {
...
       payload: {
         task: taskInfo(task),
       },
     });
     if (task.status.phase === "completed" || task.status.phase === "failed") {
-      await this.queue.publishWebhook("events.task.finished", {
+      let task_event: EventKey = "task.succeeded";
+      if (task.status.phase === "failed") {
+        task_event = "task.failed";
+      }
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: **nit:** Just saving some lines
```suggestion
      let task_event: EventKey = `task.${task.status.phase}`; // after changing the webhook to task.completed
```

File: packages/api/src/task/scheduler.ts [hunk 9]
```diff
@@ -244,22 +246,27 @@ export default class TaskScheduler {
...
-      event: "task.status",
+      event: "task.updated",
       userId: task.userId,
       payload: {
         task: taskInfo(task),
       },
     });
     if (task.status.phase === "completed" || task.status.phase === "failed") {
-      await this.queue.publishWebhook("events.task.finished", {
+      let task_event: EventKey = "task.succeeded";
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: Please use `camelCase` for any variable names as we should keep a consistent style in the code.

We should have a linter for this, but the standard in the project is `camelCase` everywhere unless maybe some static `UPPERCASE_CONSTANTS` in the root of the file.

File: packages/api/src/task/scheduler.ts [hunk 10]
```diff
@@ -285,11 +292,11 @@ export default class TaskScheduler {
       ...updates,
     };
     const timestamp = asset.status.updatedAt;
-    await this.queue.publishWebhook("events.asset.status", {
+    await this.queue.publishWebhook("events.asset.updated", {
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts:337]: We should remove the `if (!updates.status || updates.status === asset.status) {` above since we should always be sending these updates. Not only when there is a status change.

Or rather... Maybe we should be moving this `publishWebhook` up, to before the `if` check. The other webhooks below we still want to fire only when the status does change.

File: packages/api/src/task/scheduler.ts [hunk 11]
```diff
@@ -298,5 +305,23 @@ export default class TaskScheduler {
         },
       },
     });
+    if (asset.status.phase == "ready" || asset.status.phase == "failed") {
+      let asset_event: EventKey = "asset.ready";
+      if (asset.status.phase === "failed") {
+        asset_event = "asset.failed";
+      }
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: Same regarding shortening this! If you get some typing issues I think you can do this:
```suggestion
      const assetEvent = `asset.${asset.status.phases}` as EventKey;
```

Or rather, we don't even need the intermediate variable anymore in either of these, can just declare the routing key:
```ts
let routingKey = `events.asset.${asset.status.phase}` as RoutingKey;
```

File: packages/www/docs/api-reference/vod/upload.mdx [hunk 12]
```diff
@@ -83,26 +83,15 @@ axios
...
 
-You can use the returned `task` object to monitor the progress of processing the
-asset after the contents have been uploaded. Check the
-[Retrieve a Task](/docs/api-reference/vod/list-tasks#retrieve-a-task) API for
-more information.
+You can use the returned `asset` object to monitor the status of the asset after
+the contents have been uploaded. Check the
+[Retrieve a Asset](/docs/api-reference/vod/list#retrieve-an-asset) API for more
+information. You can also use the returned `task.id` to track the partial
+progress using the Task API.
```
Comment (victorges) [inline packages/www/docs/api-reference/vod/upload.mdx]: Do we have any docs on the task APIs? If so, should link them here as well.

If not, we should have that at some point 😬

File: packages/www/docs/guides/start-live-streaming/webhook.mdx [hunk 13]
```diff
@@ -102,6 +102,40 @@ this account.
...
+      description:
+        "This fires when a task is spawned. (For example, an On Demand upload)",
+    },
+    {
+      name: "task.updated",
+      description: "This fires when a task is updated.",
+    },
+    {
+      name: "task.succeeded",
+      description: "This fires when a task has succeeded.",
```
Comment (victorges) [inline packages/www/docs/guides/start-live-streaming/webhook.mdx]:
```suggestion
      name: "task.completed",
      description: "This fires when a task completes its execution successfully.",
```

File: packages/www/docs/guides/upload-a-video/index.mdx [hunk 14]
```diff
@@ -296,56 +283,12 @@ const response = await axios
...
-    ],
-    "duration": 12.031667
-  },
-  "playbackId": "4c21k7z2tjcdg8wt",
-  "playbackUrl": "https://livepeercdn.com/recordings/54aeff83-404c-490a-9afa-99296a932401/index.m3u8",
-  "downloadUrl": "https://livepeercdn.com/asset/4c21k7z2tjcdg8wt/video"
-}
-```
+- If you want to track the status of your upload, you can use the `asset`
+  endpoint by passing the `asset.id` you received during the previous request.
```
Comment (victorges) [inline packages/www/docs/guides/upload-a-video/index.mdx]:
```suggestion
  endpoint by passing the `asset.id` you received during the Request Upload request.
```

File: packages/www/docs/guides/upload-a-video/index.mdx [hunk 15]
```diff
@@ -296,56 +283,12 @@ const response = await axios
...
+  You can also track partial progress using the Task API passing the `task.id`.
```
Comment (victorges) [inline packages/www/docs/guides/upload-a-video/index.mdx]:
```suggestion
  You can also track partial progress using the Task API using the `task.id`.
```

**Review by victorges — APPROVED** [review]
LGTM with a few changes, but most are nits (which are always optional)

File: packages/www/docs/api-reference/vod/list.mdx [hunk 16]
```diff
@@ -105,29 +105,13 @@ axios
...
-                "sampleRate": 44100
-            }
-        ],
         "duration": 309.986393
     },
     "playbackId": "$PLAYBACK_ID",
     "downloadUrl": "https://livepeercdn.com/asset/$PLAYBACK_ID/video"
 }
 ```
+
```
Comment (Shih-Yu) [inline packages/www/docs/api-reference/vod/list.mdx]: The  docs for this page is now here: https://github.com/livepeer/livepeer-studio-docs/blob/main/docs/references/vod/list.mdx
I can make the changes there and @gioelecerati as a reviewer

File: packages/www/docs/api-reference/vod/upload.mdx (hunk 12 above)
Comment (Shih-Yu) [reply packages/www/docs/api-reference/vod/upload.mdx]: This is the docs for what we have now: https://github.com/livepeer/livepeer-studio-docs/blob/main/docs/references/vod/list-tasks.mdx

I can add these changes there as well 😀

File: packages/www/docs/guides/start-live-streaming/webhook.mdx [hunk 17]
```diff
@@ -102,6 +102,40 @@ this account.
       description:
         "This fires when we are no longer sending video to the multistream target.",
     },
+    {
```
Comment (Shih-Yu) [inline packages/www/docs/guides/start-live-streaming/webhook.mdx]: I will update the "new" webhook page with this information:https://github.com/livepeer/livepeer-studio-docs/blob/main/docs/guides/livestreaming/webhook.mdx

File: packages/www/docs/guides/upload-a-video/index.mdx (hunk 14 above)
Comment (Shih-Yu) [reply packages/www/docs/guides/upload-a-video/index.mdx]: I will make changes to docs: https://github.com/livepeer/livepeer-studio-docs/blob/main/docs/guides/on-demand.mdx

**Review by Shih-Yu — APPROVED** [review]

File: packages/api/src/controllers/asset.ts (hunk 1 above)
Comment (gioelecerati) [reply packages/api/src/controllers/asset.ts]: Much better!

File: packages/www/docs/guides/upload-a-video/index.mdx (hunk 14 above)
Comment (gioelecerati) [reply packages/www/docs/guides/upload-a-video/index.mdx]: While making the `tus` docs that everything was moved already on the other repo in production, so I should totally remove these docs edits from here?

File: packages/api/src/task/scheduler.ts [hunk 18]
```diff
@@ -244,22 +246,24 @@ export default class TaskScheduler {
...
-      event: "task.status",
+      event: "task.updated",
       userId: task.userId,
       payload: {
         task: taskInfo(task),
       },
     });
     if (task.status.phase === "completed" || task.status.phase === "failed") {
-      await this.queue.publishWebhook("events.task.finished", {
+      let taskEvent: EventKey = `task.${task.status.phase}`; // after changing the webhook to task.completed
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: The comment in the code was only a notice that the change would only make sense after changing the webhook to task.completed. Shouldn't include it here 😆 
```suggestion
      let taskEvent: EventKey = `task.${task.status.phase}`;
```

File: packages/api/src/controllers/asset.ts (hunk 3 above)
Comment (victorges) [reply packages/api/src/controllers/asset.ts]: Bump! Do you think this would take too much time for you? If so, I think it's ok like this and I can move it to `mung` on https://github.com/livepeer/studio/pull/1178 where I'm already doing some changes here (then you can review the PR 👼)

File: packages/api/src/controllers/asset.ts [hunk 19]
```diff
@@ -343,7 +361,7 @@ app.get("/:id", authorizer({}), async (req, res) => {
     );
   }
 
-  res.json(asset);
+  res.json(db.asset.cleanWriteOnlyResponse(asset));
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: This change should be reverted as I explained in a previous comment. The `mung` middleware enforces this on every response.

File: packages/api/src/controllers/helpers.ts (hunk 5 above)
Comment (victorges) [reply packages/api/src/controllers/helpers.ts]: Bump! Also think this should be reverted as it's only used in `user` controller. We shouldn't be using this anywhere else. Should use `writeOnly` properties for "private" fields instead.

File: packages/api/src/schema/schema.yaml (hunk 6 above)
Comment (victorges) [reply packages/api/src/schema/schema.yaml:217]: Still missing this.

File: packages/api/src/task/scheduler.ts [hunk 20]
```diff
@@ -298,5 +299,23 @@ export default class TaskScheduler {
         },
       },
     });
+    if (!updates.status || updates.status === asset.status) {
+      return;
+    }
+    if (asset.status.phase == "ready" || asset.status.phase == "failed") {
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: **humanlint:** Always use triple equals `===` in JavaScript unless you know you need double equals `==`. (oh javascript... 😆)

```suggestion
    if (asset.status.phase === "ready" || asset.status.phase === "failed") {
```

File: packages/www/docs/api-reference/vod/list.mdx [hunk 21]
```diff
@@ -105,29 +105,13 @@ axios
     "createdAt": 1644546528636,
     "videoSpec": {
         "format": "mp4",
```
Comment (victorges) [inline packages/www/docs/api-reference/vod/list.mdx]: Should probably revert these changes to docs so you're able to rebase (tons of conflicts).

**Review by victorges — CHANGES_REQUESTED** [review]
Some required, functional, changes are still pending. 
Please fix and let me know for a new review.

File: packages/www/docs/guides/upload-a-video/index.mdx (hunk 14 above)
Comment (Shih-Yu) [reply packages/www/docs/guides/upload-a-video/index.mdx]: Yes, I have the changes here for review: https://github.com/livepeer/livepeer-studio-docs/pull/158

File: packages/api/src/schema/schema.yaml (hunk 6 above)
Comment (gioelecerati) [reply packages/api/src/schema/schema.yaml:217]: Added!

File: packages/api/src/controllers/helpers.ts (hunk 5 above)
Comment (gioelecerati) [reply packages/api/src/controllers/helpers.ts]: Moved it back to `user.ts`! I moved it while making the simplified response for the `asset` then forgot it there, thanks!

File: packages/api/src/controllers/asset.ts (hunk 3 above)
Comment (gioelecerati) [reply packages/api/src/controllers/asset.ts]: Moved it into the `mung` middleware! It works, not sure it's the best / right way to do that

File: packages/www/docs/api-reference/vod/list.mdx (hunk 21 above)
Comment (gioelecerati) [reply packages/www/docs/api-reference/vod/list.mdx]: Reverted all the docs changes!

File: packages/api/src/controllers/asset.ts [hunk 22]
```diff
@@ -229,22 +229,27 @@ app.use(
...
       throw new InternalServerError("Ingest not configured");
     }
+    const query = toStringValues(req.query);
     const ingest = ingests[0].base;
     const toExternalAsset = (a: WithID<Asset>) =>
       req.user.admin
         ? withPlaybackUrls(ingest, a)
         : db.asset.cleanWriteOnlyResponse(withPlaybackUrls(ingest, a));
 
+    const cleanedAsset = (a: WithID<Asset>) =>
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: Nice!

File: packages/api/src/task/scheduler.ts [hunk 23]
```diff
@@ -273,6 +273,27 @@ export default class TaskScheduler {
...
+  async deleteAsset(asset: string | Asset) {
+    if (typeof asset === "string") {
+      asset = await db.asset.get(asset);
+    }
+    await db.asset.delete(asset.id);
+    const timestamp = asset.status.updatedAt;
+    await this.queue.publishWebhook("events.asset.deleted", {
+      type: "webhook_event",
+      id: uuid(),
+      timestamp,
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: The webhook timestamp should be the time on which the respective event happened, and the asset `updatedAt` is not (and doesn't need to) be updated when it is deleted.

IOW:
```suggestion
    await this.queue.publishWebhook("events.asset.deleted", {
      type: "webhook_event",
      id: uuid(),
      timestamp: Date.now(),
```

**Review by victorges — APPROVED** [review]
LGTM! :shipit:

File: packages/api/src/task/scheduler.ts (hunk 23 above)
Comment (gioelecerati) [reply packages/api/src/task/scheduler.ts]: Right! makes sense

**Comment by victorges** [comment]
lets merge this

**Comment by victorges** [comment]
@gioelecerati after merging some stuff to `master` that added conflicts on this, I've just rebased this one to `master` myself. Included 2 changes apart from that: 
 - Fixed a regression that was calling `db.asset.delete` instead of `markDeleted`: https://github.com/livepeer/studio/pull/1154/commits/bc5686442c42e1ca020f3882191683027200335f#diff-72488d88e2bc58582e64d3c8e8283bcfb59c422a35fcf476d51a2e1da9e41716R305
 - Improve the `mung` middleware changes (I had refactored it in another PR so there was a much clearer integration path there). Also made `details` automatically `true` for `admins` 😎 https://github.com/livepeer/studio/pull/1154/commits/47738a927e2198bda3dcf9cbb653360a14890224 
 - Also a fix on the `updateAsset` func as it would always send a webhook for "ready" or "failed" assets. It should only do so when there is actually a change in the `phase`: https://github.com/livepeer/studio/pull/1154/commits/a7aa7cc4bfe286c08d07b63d13a2c0a25f58ce75

**Review by victorges — APPROVED** [review]
LGTM! Let's launch this

**Comment by victorges** [comment]
👀
